### DISCUSSION TITLE
Can we hit or use the meshcentral's websocket externally from other application.

### DISCUSSION THREAD

**Discussion by a-imran99** [Ylianst/MeshCentral#4292] [Q&A]
So I can elaborate here. 
The idea is to use Meshcentral websocket which **runs as control.ashx from another application** where **meshcentral is embedded in Iframe** ( a parent application which has meshcentral embedded to it ). The idea is to change machine/systems info but not from the general info screen provided by meshcentral but from another place in parent application.

So I explored that when changing consent
```
{
  action: "changedevice"
  consent: 7
  nodeid: "node//iPoJtq84VWOBjS5WDMbl18UV5........"
}
```

The above action request is sent to websocket to change the consent of the particular node.

Is there a way to achieve this externally. another point to say is, **the parent application in hosted on www.example.com and meshcentral is hosted in subdomain as www.example.com/mesh** .

**Comment by VINISHVV** [comment]
Please refer [here](https://github.com/Ylianst/MeshCentral/issues/2533#issuecomment-824472025)

**Reply by a-imran99** [reply]
Hi @VINISHVV, thanks for the reply.

Embedding meshcentral in iframe is not the question, but using meshcentral's websocket is the doubt. How can we hit meshcentral's websocket using wscat, websocat or postman outside the application.

@Ylianst It would be great if you made some comment. Thank you!

**Comment by Ylianst** [comment]
Yes, you can. First, if you are curious what flows over the `control.ashx` websocket, you can add "?trace=1" to the URL of the main MeshCentral web page and open the browser console. You will see all the JSON messages going both directions.

![image](https://user-images.githubusercontent.com/1319013/180136041-6f68e291-e0df-43c4-aa44-22d263a7eac0.png)

Once you take a look at the messages, it's not too difficult to write your own JavaScript application that does the same. There is another place you should at, and that is the "meshctrl.js" code. This is a command line tool that comes with MeshCentral, you run it like this:

```
node node_modules/meshcentral/meshctrl.js
```

It will show something like this:

```
MeshCtrl performs command line actions on a MeshCentral server.
Information at: https://meshcentral.com
No action specified, use MeshCtrl like this:

  meshctrl [action] [arguments]

Supported actions:
  Help [action]               - Get help on an action.
  ServerInfo                  - Show server information.
  UserInfo                    - Show user information.
  ListUsers                   - List user accounts.
  ListUserSessions            - List online users.
(...)
  DeviceToast                 - Display a toast notification on a remote device.
  DevicePower                 - Perform wake/sleep/reset/off operations on remote devices.
  DeviceSharing               - View, add and remove sharing links for a given device.
  AgentDownload               - Download an agent of a specific type for a device group.

Supported login arguments:
  --url [wss://server]        - Server url, wss://localhost:443 is default.
                              - Use wss://localhost:443?key=xxx if login key is required.
  --loginuser [username]      - Login username, admin is default.
  --loginpass [password]      - Login password.
  --token [number]            - 2nd factor authentication token.
  --loginkey [hex]            - Server login key in hex.
  --loginkeyfile [file]       - File containing server login key in hex.
  --logindomain [domainid]    - Domain id, default is empty, only used with loginkey.
  --proxy [http://proxy:123]  - Specify an HTTP proxy.
```

If you look at the code, you will see that meshctrl.js opens a websocket to `control.ashx` and does the same commands as the main web page. MeshCentral Router also opens a web socket to `control.ashx`, if you are interested in writing C# code that does this.

In any case, my recommendation is to look at meshctrl.js as a starting point. You can probably write some crazy powerful application to various purposes.

The one thing is how to authenticate to `control.ashx`. There are many ways to do it. You can use a username/password in the header, session cookies, etc. So, depending on what you are trying to do, this will vary. One easy way to get started is to just create a new "myapp.htm" in the `node_modules\meshcentral\public` folder. You can then login to MeshCentral and browse to that page and open a websocket connection.

Ok, hopefully that was useful. Let me know if I can help more.

**Reply by a-imran99** [reply]
@Ylianst Thank you . got it. Thanks for the quick answer. 
The parent application in on react and I needed to access and authenticate the control.ashx from the react application. Is that possible?

Thanks for the reply.

**Reply by Ylianst** [reply]
Sure. When making your websocket request, add the following header:

```
x-meshauth: [usernameInBase64],[passwordInBase64]
```

So, it should look like:

```
x-meshauth: dXNlcg==,UGFzc3dvcmQ=
```

That will get your logged in and connected. If you have a 2FA token, add it as a 3rd Base64 like this:

```
x-meshauth: dXNlcg==,UGFzc3dvcmQ=,MTIzNDU2
```
